gcc/rust/backend/rust-compile-type.cc

```diff
@@ -239,14 +241,54 @@ TyTyResolveCompile::visit (const TyTy::FnPtr &type)
 					   type.get_ident ().locus);
 }
 
+void
+check_variant_record_collision (Context *ctx, const TyTy::ADTType &type,
+				std::vector<tree> &variant_records)
+{
+  // bdbt: we're checking if shared discriminants crash with each other or
+  // not. lets make a map from uhwi to hir id. A clash of uhwi in a variant
+  // record to which said record can be converted uhwi is indicative of
+  // issue 3351 of gccrs
+
+  size_t i = 0;
+  std::map<HOST_WIDE_INT, std::vector<size_t>> shwi_to_index;
+  for (; i < variant_records.size (); i++)
+    {
+      TyTy::VariantDef *variant = type.get_variants ().at (i);
+      if (variant->has_discriminant ())
+	{
+	  tree discriminant_expr
+	    = CompileExpr::Compile (variant->get_discriminant (), ctx);
+	  HOST_WIDE_INT discriminant_integer
+	    = tree_to_shwi (fold_expr (discriminant_expr));
+	  shwi_to_index[discriminant_integer].push_back (i);
+	}
+    }
+
+  for (auto &[discriminant_integer, index_vector] : shwi_to_index)
+    {
+      // collision happens
+      if (index_vector.size () >= 2)
```

Comment:
i would probably write this as:

```
bool collision = iindex_vector.size () > 1;
if (!collision)
   continue;
```

It will flatten the code make it bit neater.